Yoast/wpseo-news
# PHPCS: Update to YoastCS 1.3.0

## Summary
This PR can be summarized in the following changelog entry:
* _N/A_

## Relevant technical choices:

Note: As the News plugin was so far behind, the update to YoastCS 2.0 is done in two steps - first to YoastCS 1.3.0 (this PR), then to YoastCS 2.0.
This allows for slightly easier reviewing of the changes.

### Composer: up the requirement for YoastCS

* A selective update of the `composer.lock` for just YoastCS and its dependencies.
* The `installed_paths` command is now no longer needed. This will be handled automatically by the Composer plugin which comes with YoastCS 1.2.0+.
    To that end, remove the `composer config-yoastcs` command from the Travis script.
* To not break BC for devs who are used to having this command available, update the composer script command to call the Dealerdirect plugin to handle this.

### Build/PHPCS: Rename PHPCS configuration file

Using a `.dist` extension allows for individual developers to overload the `phpcs.xml.dist` with a local `phpcs.xml` file. This file can include the `dist` file as a base ruleset using `<rule ref="./.phpcs.xml.dist"/>`.

To facilitate this, the typical file names for the overloaded PHPCS config file have been added to the `.gitignore` file.

### Build/PHPCS: set up the PHPCS ruleset properly

* Add the XSD.
* Add fencing and documentation.
* Provide all command line arguments.
* Remove the exclusion of the `tests` directory.
* Remove the exclusion of the dependency related directories. These are excluded by default since YoastCS 1.3.0.
* Remove sniff `exclude`s which have clearly been copy/pasted from the original WPSEO ruleset.
* Don't hide PHPCS warnings as a default.
* Use tab based indentation in the ruleset file in line with indentation elsewhere.

### Build/PHPCS: list the custom test classes

Several sniffs will make exceptions for code in unit test classes.
This is determined by checking whether these classes extend known unit test case classes.

This property provides those class names to the sniffs.

### Build/PHPCS: specify the text-domain which should be used

... to enable the check for this within the WP.I18n sniff.

Includes fixing the `Text domain` header in the main plugin file which was using an incorrect text domain.

### Build/PHPCS: add configuration for the `Yoast.Files.FileName` sniff

### Build/PHPCS: enable checking for globals being prefixed

### Build/PHPCS: add configuration for the `Yoast.Files.TestDoubles` sniff

### Build/PHPCS: add a few selective exclusions

... includes documentation on why these files should be excluded for these specific rules.

* Allow for overloading method just to change visibility in test situations.
* Allow non-prefixed globals in select test situations.
* Allow access to and overloading of superglobals in the unit test case file.
* Allow using `@inheritdoc` for methods in test double classes.

### Build/PHPCS: temporary tweak for short array format

WordPressCS will demand long arrays and forbid the use of short arrays as of WPCS 2.2.0.
See: https://make.wordpress.org/core/2019/07/12/php-coding-standards-changes/

For YoastCS, however, a choice has been made to demand short arrays.

As support for PHP < 5.6 has been dropped, this can be enforced everywhere now and will be as of YoastCS 2.0.0 which is expected soon.

For now, just ignore use of short arrays.

### Build/PHPCS: add temporary exclusion

This need further review/attention and has a `technical-debt` ticket attached to it - see #585.

### PHPCS: replace an old-style ignore with a new-style ignore

### PHPCS: update an ignore comment to make it selective

### PHPCS: whitelist usage of WPSEO hooks

### PHPCS: whitelist a direct database query

... which is only used when a new version of the plugin is installed and not as a regular page query.

### PHPCS: whitelist select use of readfile() 



## Test instructions

* Run `composer install`
* Run `composer check-cs` & see a clean result.
